eval.py

```diff
@@ -0,0 +1,217 @@
+        """
+        vec = np.zeros(len(label_set), dtype=int)
+        for i in intent.split("#"):
+            vec[label_set[i]] = 1
+        return list(vec)
+
+    def convert_slot(self, slot, label_set):
+        """ Convert slot to 0-1 vector.
+        :slot: slot in string format
+        :label_set: an ordered set of slots
+        """
+        vec = np.zeros(len(label_set), dtype=int)
+        vec[label_set[slot]] = 1
+        return list(vec)
+
+    def generate_intent_map(self, intent_list):
+        """ Generate an ordered map for all intents. """
+        intent_set = set()
+        for i in intent_list:
+            intent_set.update(set(i.split("#")))
+
+        intent_set = {i:j for j, i in enumerate(intent_set)}
+        return intent_set
+
+    def generate_slot_map(self, slot_list):
+        """ Generate an ordered map for all slots. """
+        slot_set = set()
+        for s in slot_list:
+            slot_set.update(set(s))
+
+        if self.exclude_other:
+            slot_set.remove("O")
+        slot_set = {s:i for i, s in enumerate(slot_set)}
+        return slot_set
+
+    def generate_summary(self):
+        """ Organize pred_out by query length. """
+        # summary_intent is a dictionary with query-length as keys
+        # for each query length, it contains prediction result at each incremental step
+        # each element in the list corresponds to a (golden, pred) pair 
+        self.summary_intent = defaultdict(lambda: defaultdict(lambda: [[], []]))
+        self.summary_slot = defaultdict(lambda: defaultdict(lambda: [[], []]))
+        self.seq = defaultdict(lambda: defaultdict(list)) 
+
+        t_b = 0
+        t_a = 0
+        for sent_id, rec in self.test_set.items():
+            query_len = max(rec.keys())
+
+            for num_token in rec:
+                y_true = self.convert_intent(rec[num_token]["gold_int"], self.intent_set)
+                y_pred = self.convert_intent(rec[num_token]["pred_int"], self.intent_set)
+                self.summary_intent[query_len][num_token][0].append(y_true)
+                self.summary_intent[query_len][num_token][1].append(y_pred)
+
+                slots = np.array(rec[num_token]["gold_slot"][1:])
+                filter_o = np.where(slots != "O")[0]
+                filtered_slots = slots[filter_o]
+                t_b += len(slots)
+                t_a += len(filtered_slots)
+                s_true = [self.convert_slot(s, self.slot_set) for s in filtered_slots]
+                s_pred = [self.convert_slot(s, self.slot_set) for s in filtered_slots]
+                self.summary_slot[query_len][num_token][0].extend(s_true)
+                self.summary_slot[query_len][num_token][1].extend(s_pred)
+
+                self.seq[query_len][num_token].extend([sent_id] * len(filtered_slots))
+
+        self.summary = {"intent": self.summary_intent, "slot": self.summary_slot}
+        print(t_b, t_a)
+
+    def calculate_slot_f1(self, granularity=0.05):
```

Comment:
It might be good to have some sort of unit test for this function and the calculate_intent_f1 function, to make sure that the calculation is correct.